### PR TITLE
feat(web): start fetching data for zone at historical datetime

### PR DESCRIPTION
**Pull request by tonypls**
## Issue
Allow fetching of historical hourly data from the front end

## Description
This PR adds support for a user to visit the app via a historical link and navigate through the data for a given historical period.

Testing instructions: Run the backend: https://github.com/electricitymaps/electricitymaps/pull/5723
Visit: http://localhost:5173/zone/SE-SE2/hourly/2024-10-20T08:00:00Z

### CODE REVIEW

File: web/src/api/getState.ts [hunk 1]
```diff
@@ -2,14 +2,33 @@ import type { UseQueryResult } from '@tanstack/react-query';
...
-const getState = async (timeAverage: string): Promise<GridState> => {
-  const path: URL = new URL(`v8/state/${timeAverage}`, getBasePath());
-  path.searchParams.append('cacheKey', cacheBuster());
+const getState = async (
+  timeAverage: string,
+  targetDatetime?: string
+): Promise<GridState> => {
+  const parsedPath = parsePath(location.pathname);
+  const isValidDatetime = targetDatetime && isValidDate(targetDatetime);
+  const timeAverageToQuery = parsedPath?.timeAverage || timeAverage;
```
Comment (VIKTORVAV99) [inline web/src/api/getState.ts:20]: Should these not always be in sync?

File: web/src/api/getState.ts [hunk 2]
```diff
@@ -2,14 +2,33 @@ import type { UseQueryResult } from '@tanstack/react-query';
...
 
-const getState = async (timeAverage: string): Promise<GridState> => {
-  const path: URL = new URL(`v8/state/${timeAverage}`, getBasePath());
-  path.searchParams.append('cacheKey', cacheBuster());
+const getState = async (
+  timeAverage: string,
+  targetDatetime?: string
+): Promise<GridState> => {
+  const parsedPath = parsePath(location.pathname);
+  const isValidDatetime = targetDatetime && isValidDate(targetDatetime);
```
Comment (VIKTORVAV99) [inline web/src/api/getState.ts:19]:
```suggestion
  const isValidDatetime = Boolean(targetDatetime) && isValidDate(targetDatetime);
```
To ensure type cornecctness later on.

File: web/src/api/getState.ts [hunk 3]
```diff
@@ -2,14 +2,33 @@ import type { UseQueryResult } from '@tanstack/react-query';
...
+    `v8/state/${timeAverageToQuery}${
+      isValidDatetime ? `?targetDate=${targetDatetime}` : ''
+    }`,
+    getBasePath()
+  );
 
+  path.searchParams.append(
+    'cacheKey',
+    isValidDatetime && targetDatetime ? targetDatetime : cacheBuster()
+  );
```
Comment (VIKTORVAV99) [inline web/src/api/getState.ts:31]: I think we can skip adding this entriely if the targetDatetime is valid.

We already have the targetDate in the query parameters that will bust the cache.

**Review by VIKTORVAV99 — COMMENTED** [review]
Bunch of comments, haven't done any testing yet but will try and do so as tomorrow 🙂

File: web/src/App.tsx [hunk 4]
```diff
@@ -31,21 +36,34 @@ const TimeControllerWrapper = lazy(() => import('features/time/TimeControllerWra
...
+
+  // Set initial datetime synchronously
+  useLayoutEffect(() => {
+    if (parsedPath?.datetime) {
+      const pathDate = new Date(parsedPath.datetime);
+      if (!Number.isNaN(pathDate.getTime())) {
+        setTargetDatetimeString(parsedPath.datetime);
+        setSelectedDatetimeIndex({
+          datetime: pathDate,
+          index: 24,
```
Comment (VIKTORVAV99) [inline web/src/App.tsx:52]: I don't think we can set the index to 24 here, at least without additional checks. Currently this also makes it so we have 25 months and years of data 😅

File: web/src/api/getState.ts (hunk 3 above)
Comment (tonypls) [reply web/src/api/getState.ts:31]: Good point, it was useful to see for testing while building. I'll try remove it now

File: web/src/utils/pathUtils.test.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        timeAverage: 'hourly',
+      });
+
+      expect(parsePath('/zone/US-CA/hourly/2024-10-20T08:00:00Z')).to.deep.equal({
+        type: 'zone',
+        zoneId: 'US-CA',
+        timeAverage: 'hourly',
+        datetime: '2024-10-20T08:00:00Z',
+      });
+    });
```
[the remark this answers is not in the document]
Comment (tonypls) [reply web/src/utils/pathUtils.test.ts]: Good point, I've updated this!

File: web/src/features/map/Map.tsx [hunk 6]
```diff
@@ -253,12 +257,32 @@ export default function MapPage({ onMapLoad }: MapPageProps): ReactElement {
       setHoveredZone(null);
       if (feature?.properties) {
         const zoneId = feature.properties.zoneId;
-        navigate(createToWithState(`/zone/${zoneId}`));
+        navigate(
+          createToWithState(
+            `/zone/${zoneId}${timeAverage ? `/${timeAverage}` : ''}${
+              targetDatetime ? `/${targetDatetime}` : ''
+            }`
+          )
```
[the remark this answers is not in the document]
Comment (tonypls) [reply web/src/features/map/Map.tsx:268]: Yeah, I think so. I tried at first but it's not a hook so I had to pass it the timeaverage and targetdatetime. Ideally we hookify it. I'll try do it if I find the time!

File: web/src/api/getState.ts (hunk 1 above)
Comment (tonypls) [reply web/src/api/getState.ts:20]: Good point

**Comment by tonypls** [comment]
Hold of on detailed reviews for now, I'll separate the routing changes into it's own branch!

**Comment by tonypls** [comment]
Closing in favour of #7397

**Pull request closed without merging**